### DISCUSSION TITLE
want to Move meta tags at top of <head> for SEO, tried gatsby-ssr solution but not getting result

### DISCUSSION THREAD

**Discussion by abhaybriskteq** [gatsbyjs/gatsby#29455] [Help]
I want to move meta tags to the top of <head> components so facebook can recognize it. I used gatsby-ssr given code but cannot get expected output

**Code:**
```
const { Helmet } = require('react-helmet');

exports.onRenderBody = ({
  setHeadComponents,
  setHtmlAttributes,
  setBodyAttributes,
}) => {
  const helmet = Helmet.renderStatic();
  console.log('helmet', helmet);
  setHtmlAttributes(helmet.htmlAttributes.toComponent());
  setBodyAttributes(helmet.bodyAttributes.toComponent());
  setHeadComponents([
    helmet.meta.toComponent(),
    helmet.title.toComponent(),
    helmet.link.toComponent(),
    helmet.noscript.toComponent(),
    helmet.script.toComponent(),
    helmet.style.toComponent(),
  ]);
};

exports.onPreRenderHTML = ({ getHeadComponents, replaceHeadComponents }) => {
  const headComponents = getHeadComponents();
  const metaComponents = [];
  headComponents.map((component) => {
    if (component.type === 'meta') {
      metaComponents.push(component);
    }
    return component;
  });
  const newHeadComponents = [...metaComponents];
  headComponents.map((component) => {
    if (component.type !== 'meta') {
      newHeadComponents.push(component);
    }
    return component;
  });
  replaceHeadComponents(newHeadComponents);
};
```

**Output:**

when I get console output of onRenderBody..

```
helmet {
  base: {
    toComponent: [Function: toComponent],
    toString: [Function: toString]
  },
  bodyAttributes: {
    toComponent: [Function: toComponent],
    toString: [Function: toString]
  },
  htmlAttributes: {
    toComponent: [Function: toComponent],
    toString: [Function: toString]
  },
  link: {
    toComponent: [Function: toComponent],
    toString: [Function: toString]
  },
  meta: {
    toComponent: [Function: toComponent],
    toString: [Function: toString]
  },
  noscript: {
    toComponent: [Function: toComponent],
    toString: [Function: toString]
  },
  script: {
    toComponent: [Function: toComponent],
    toString: [Function: toString]
  },
  style: {
    toComponent: [Function: toComponent],
    toString: [Function: toString]
  },
  title: {
    toComponent: [Function: toComponent],
    toString: [Function: toString]
  }
}

```

And when I hit console.log(helmet.meta.toStirng) or console.log(helmet.meta.toComponent). It shows null for all.
Can anyone help me regarding this?

**Comment by vladar** [comment]
Are you using [gatsby-plugin-react-helmet](https://www.gatsbyjs.com/plugins/gatsby-plugin-react-helmet/)? In this case, you don't need to adjust `gatsby-ssr`, the plugin does it for you. It should just work.

Also, check out related docs: https://www.gatsbyjs.com/docs/how-to/adding-common-features/seo/#page-metadata

**Reply by abhaybriskteq** [reply]
@vladar I tried that also. but not getting any result.

**Reply by patelvarun** [reply]
I am also facing the same issue.  gatsby-plugin-react-helmet adding metadata at the end of head tag. So all Social media are not capturing metadata.

**Reply by patelvarun** [reply]
See Head tag. In the attached image see the gatsby-plugin-react-helmet adding meta tag at last.
<img width="1536" alt="Screenshot 2021-02-25 at 6 38 52 PM" src="https://user-images.githubusercontent.com/6904686/109158061-deb3ed00-7798-11eb-9b8b-73ad5296ea57.png">

**Reply by patelvarun** [reply]
@vladar - Can you please guide me on how we can move metadata on top of the head section?

**Comment by LukaszW007** [comment]
I have exactly the same problem. I have deployed a gatsby generated website with blog posts generated via gatsby-node.js. All of the pages contain all the required metatags - I use Helmet. The main page is crawled by facebook properly so when I want to share a link to it in an FB post, the link contains the correct title, link, and thumbnail. Unfortunately, blog post pages even though have proper metatags are crawled partly by FB, like Facebook would crawl only several first lines in a HEAD tag until plenty of LINK. My HTML head looks exactly the same as @patelvarun - 4 meta tags on the very top of head, many lin> and script tags and on the very end - meta tags again with og:image, og:title etc.
I believe that moving <meta> tags from the bottom of the list of tags in <head> to the very top would fix the improper facebook crawling issue.
Does anyone know how to do it?

**Reply by LukaszW007** [reply]
![obraz](https://user-images.githubusercontent.com/19213851/121819970-62361300-cc90-11eb-8f8c-c37113f94262.png)

**Comment by askibinski** [comment]
Apparantly `react-helmet` does not support inserting specific tags at a specific place: https://github.com/nfl/react-helmet/issues/514

So, if you need that, you may have to do it manually using `useEffect`, see issue for example. But that seems like a poor solution There might be better alternatives though?

**Reply by LekoArts** [reply]
This should be fixed with https://github.com/gatsbyjs/gatsby/pull/34030 which will be released in 4.5.0 today

**Reply by askibinski** [reply]
Ah amazing! Thanks!
